Approximate comparison (re: issue 38 - https://github.com/facebook/ios-snapshot-test-case/issues/38)

nap-sam-dean opened the pull request:
Hey.

I've added in an option to pass a maxumim number of pixels that can differ for an image to be counted as equal.

This solves my particular use case (we have an errant pixel in a gradient that's throwing our tests off) but if people want more control this could be extended to include different metrics of differenceness.

Let me know what you think!

nap-sam-dean commented:
PS I don't know of any standards for pull requests - I've made tests for my changes, the travis build script passes and the pod lints. I didn't update the podspec version number because I don't know what else goes into releases etc. I also didn't squash my commits, but I can if you need me to. S

dstnbrkr commented:
Thanks @nap-sam-dean! Will take a closer look once we get a CLA on file. Here's some initial feedback:
- can we express max difference as a percent?
- instead of adding a new comparison method, can we just add a parameter to the old?
- looks like the project file and podspec got churned a bit because of the new dependency, but since that dependency isn't necessary those files don't need to be changed in this PR.

nscoding commented:
This would be such a great addition.

* Yes I agree with @dstnbrkr, we shouldn't have a new comparison method but this will make it a breaking change though.
* Don't forget to have the same changes for `FBSnapshotVerifyLayer` macros

nap-sam-dean commented:
If you don't mind breaking changes then that would certainly be cleaner :) 

@dstnbrkr When you say 'add a parameter to the old' do you mean from the #define onwards?

And I'm waiting for net-a-porter to get back to me about the cla, sigh. I should have done this in my own time :)

nap-sam-dean commented:
Hi.

I've updated the pull request to include

- percentage difference instead of absolute difference
- removed annoying changes in the project file
- just updated the original macro / methods instead of adding additional ones
- updated the ..Layer macros, sorry @nscoding !
- reverted the Podfile.lock - this was changed because of the `pod install` step in `build.sh` - looks like I'm using a slightly different version of the cocoapods gem!

More feedback welcome, I'm still waiting on the HR team here so I can sign the SLA :(

dstnbrkr commented on FBSnapshotTestCase/FBSnapshotTestCase.h:
```diff
@@ -50,24 +51,25 @@ FBSnapshotVerifyViewWithOptions(view__, identifier__, FBSnapshotTestCaseDefaultS
 }
 
-#define FBSnapshotVerifyLayer(layer__, identifier__) \
+#define FBSnapshotVerifyLayer(layer__, identifier__, tolerance__) \
```
Sorry to not suggest this earlier: since FBSnapshotVerify(View/Layer)WithOptions [was added so recently](https://github.com/facebook/ios-snapshot-test-case/commit/d0a4848be03bb1c938e415b3e426b43614927980#diff-00c0098c680647a84df94272d707271eR30), I think we can churn that one and leave FBSnapshotVerify(View/Layer) unchanged. FBSnapshotVerify(View/Layer)  can call FBSnapshotVerify(View/Layer)WithOptions with a 0 tolerance to preserve the current behavior. I think you had a default value in the previous rev.

nscoding replied:
Right, we should keep this the same and only append a parameter on `FBSnapshotVerifyLayerWithOptions` (like is done already)

dstnbrkr commented on FBSnapshotTestCase/FBSnapshotTestCase.m:
```diff
@@ -68,12 +72,14 @@ - (BOOL)compareSnapshotOfView:(UIView *)view
   return [_snapshotController compareSnapshotOfViewOrLayer:viewOrLayer
                                                   selector:self.invocation.selector
                                                 identifier:identifier
+                                        tolerance:tolerance
```
nit: whitespace

dstnbrkr commented on FBSnapshotTestCase/FBSnapshotTestController.m:
```diff
@@ -282,33 +285,36 @@ - (BOOL)compareSnapshotOfView:(UIView *)view
 #pragma mark Private API
 
-- (BOOL)_performPixelComparisonWithViewOrLayer:(UIView *)viewOrLayer
+- (BOOL)_performPixelComparisonWithViewOrLayer:(id)viewOrLayer
```
yeesh, good catch

nscoding commented on FBSnapshotTestCase/FBSnapshotTestCase.m:
```diff
@@ -43,22 +43,26 @@ - (void)setRecordMode:(BOOL)recordMode
   return [self _compareSnapshotOfViewOrLayer:layer
                     referenceImagesDirectory:referenceImagesDirectory
                                   identifier:identifier
+                          tolerance:tolerance
```
nit: spacing

nscoding commented:
Overall looks awesome, it will be such a great addition to the framework.

nap-sam-dean commented:
Hey.

Latest changes:

- Remove both silly spacing mistakes in `FBSnapshoptTestCase.m`
- Remove the tolerance parameter for the basic macro; t now assumes a tolerance of 0

dstnbrkr commented:
Looks like FBSnapshotTestCaseDemoTests still needs an update (still passes third arg to FBSnapshotVerifyView).

nap-sam-dean commented:
Well, that's pretty embarrassing :) Could have sworn I ran `./build.sh` before I pushed. Oh well, it should pass now!

nap-sam-dean commented:
Hi everyone.

I've signed the CLA and pushed fixes to the conflicts. However, I've got a few questions: 

1) For some reason my tests fail if we scale the context on retina devices. Looking at the pixel memory without scaling I get the correct values but with scaling I just see 0xFF indicating that the image isn't being drawn correctly. However, I don't know why we need to scale the context so I don't know what effect removing the scaling code has (I suspect something subtle will break). Instead of using the scaling value for the screen should we be using the scale property of the incoming image? (And should we put a check in place so that images with different scale values aren't equal?)

2) The CI is failing because it can't find an example image - should I run the failing test in recode mode and commit the reference image for this failing test? I can do that no problem but I'm surprised that it's been passing up until now :)


Thanks everyone :)

dstnbrkr commented:
Thanks @nap-sam-dean! The most recent merge changed the reference image directory configuration from a preprocessor macro to an environment variable. I see in the Travis CI logs that the reference image directory isn't set correctly:

FBReferenceImageFilePathKey=P‹_32/FBSnapshotTestCaseDemoTests/testExample@2x.png

That "P‹" before the "_32" looks like a bug - but no idea how that bug would be first appearing in this PR. I'll investigate as well, but if you want to take a look - the problem might be in how the path to the reference image is constructed.

dstnbrkr commented on FBSnapshotTestCase/FBSnapshotTestCase.h:
```diff
@@ -30,8 +31,8 @@ BOOL comparisonSuccess__; \
 for (NSString *suffix__ in suffixes__) { \
-NSString *referenceImagesDirectory__ = [NSString stringWithFormat:@"%@%@", envReferenceImageDirectory, suffix__]; \
-comparisonSuccess__ = [self compareSnapshotOfView:(view__) referenceImagesDirectory:referenceImagesDirectory__ identifier:(identifier__) error:&error__]; \
+NSString *referenceImagesDirectory__ = [NSString stringWithFormat:@"%s%@", envReferenceImageDirectory, suffix__]; \
```
Ah, here it is - that %s needs to be a %@ (explains why the reference image path looks like a random string)

nscoding replied:
Ah nice, let's fix this and travis should be green.

dstnbrkr commented:
We don't have a CLA on file yet @nap-sam-dean, can you submit once more? @facebook-github-bot will comment on this PR once it receives the CLA.

dstnbrkr commented on FBSnapshotTestCase/UIImage+Compare.m:
```diff
@@ -76,15 +84,53 @@ - (BOOL)compareWithImage:(UIImage *)image
+  CGFloat scaleFactor = [[UIScreen mainScreen] scale];
+  CGContextScaleCTM(referenceImageContext, scaleFactor, scaleFactor);
+  CGContextScaleCTM(imageContext, scaleFactor, scaleFactor);
+    */
```
My guess is that scaling changes the number of pixels that the algorithm below needs to enumerate.

nap-sam-dean replied:
I thought it might be that so I tested it by setting the very first pixel in my test be the one that should be different. I got white for both images when scaled to 2x, white vs. black when scaled to 1x (by testing on the iPhone vs. iPad)

nscoding replied:
Right, we don't need to scale this here since `CGImageGetWidth()` and `CGImageGetHeight()` will return the right value. I wonder why we had this before?

nscoding commented on FBSnapshotTestCase/UIImage+Compare.m:
```diff
@@ -76,15 +84,53 @@ - (BOOL)compareWithImage:(UIImage *)image
+      }
+
+      p1 ++;
+      p2 ++;
```
nit: Remove the space between `p1/2` and `++`

dstnbrkr commented:
ping! couple minor test fixes and CLA resubmit and this PR is all set

nap-sam-dean commented:
Hi,

Sorry for the small pause :)

I've signed the CLA again (my mistake, I'd signed as my private github account, not my net-a-porter one, sigh)

I've fixed the reference images path, removed the commented out scaling code and removed the spaces before the `++`s. The CI build is passing again now :)

PS The issue was a bad merge by me to resolve the conflict from pull request 89 :(

dstnbrkr commented:
Thanks for this great contribution @nap-sam-dean! I'll defer to @nscoding for the merge.

nscoding commented:
Thank you so much @nap-sam-dean

nap-sam-dean commented:
No problem :)

kerrmarin commented:
:clap: :clap: 

Is it possible, now that this is merged, to get a release that includes this feature?

Thanks!

nscoding commented:
@kerrmarin Yes of course, we will let it sit on master for few days and then we will tag it as 2.0
